Block volumes Support: CRI, volumemanager and operationexecutor changes

mtanino opened the pull request:
**What this PR does / why we need it**:
This PR contains following items to enable block volumes support feature.
- container runtime interface change
- volumemanager changes
- operationexecuto changes


**Which issue this PR fixes**:
Based on this proposal (kubernetes/community#805) and this feature issue: kubernetes/features#351

**Special notes for your reviewer**:

There are another PRs related to this functionality.
(#50457) API Change
(#53385) VolumeMode PV-PVC Binding change
(#51494) Container runtime interface change, volumemanager changes, operationexecutor changes
(#55112) Block volume: Command line printer update
Plugins
(#51493) Block volumes Support: FC plugin update
(#54752) Block volumes Support: iSCSI plugin update

**Release note**:
```
Adds alpha support for block volume, which allows the users to attach raw block volume to their pod without filesystem on top of the volume.
```
/cc @msau42 @liggitt @jsafrane @saad-ali @erinboyd @screeley44

msau42 commented on pkg/volume/util/operationexecutor/operation_generator.go:
```diff
@@ -579,49 +588,241 @@ func (og *operationGenerator) GenerateUnmountDeviceFunc(
+		return nil, "", volumeToMount.GenerateErrorDetailed("MapVolume. Find PersistentVolume failed. PersistentVolume is nil.", nil)
+	}
+	claimName := pv.Spec.ClaimRef.Name
+	pvc, _ := og.kubeClient.Core().PersistentVolumeClaims(volumeToMount.Pod.ObjectMeta.Namespace).Get(claimName, metav1.GetOptions{})
```
Can you use the volume mode from the PV instead of making an API call to get the PVC?

mtanino replied:
Ack.
Based on the latest bind matrix, we can decide VolumeMode by only referring the PV volumeMode after passing the pv-binder.

msau42 commented on pkg/volume/util/operationexecutor/operation_generator.go:
```diff
@@ -579,49 +588,241 @@ func (og *operationGenerator) GenerateUnmountDeviceFunc(
-			deviceOpened, deviceOpenedErr = mounter.DeviceOpened(deviceToDetach.DevicePath)
-			if deviceOpenedErr != nil {
-				return deviceToDetach.GenerateErrorDetailed("UnmountDevice.DeviceOpened failed", deviceOpenedErr)
+		// Try to unmap symlink on deviceMountPath
```
Why are there changes to Unmount when we have a separate Unmap function?

mtanino replied:
I need to check this again whether we can get persistent volume object from AttachedVolume struct.
If we can get PV object, then we can make this method separately.

msau42 commented on pkg/volume/util/operationexecutor/operation_generator.go:
```diff
@@ -760,3 +961,85 @@ func checkNodeAffinity(og *operationGenerator, volumeToMount VolumeToMount, plug
+	return nil
+}
+
+func unmapDevice(globalPDPath string) error {
```
Shouldn't the plugin be doing these?

mtanino replied:
If we implement this method here, each plugin doesn't need to implement this method individually.
But since this method has OS dependent implementation like os.lstat, os.remove, this might better to be implemented at individual plugin.
Any idea?

msau42 replied:
If we want to follow the same model as filesystem volumes, then the plugin should do all the work.  But I do see value in having kubelet handle common methods that are used by all plugins.  At this point, do we know for sure that there are common steps that are the same for all plugins?

mtanino replied:
@msau42 
The clean up steps are under discussion with @jsafrane on the spec doc.
Could you give feedback?

https://github.com/kubernetes/community/pull/805/files#diff-b828f85c1086b7acfaf489fe8605329fR609

msau42 commented on cmd/kubelet/app/options/options.go:
```diff
@@ -39,6 +39,7 @@ import (
 const (
 	DefaultKubeletPodsDirName       = "pods"
 	DefaultKubeletVolumesDirName    = "volumes"
+	DefaultKubeletDevicesDirName    = "devices"
```
Should this be something like "volumeDevices" to distinguish it from generic devices?

mtanino replied:
ack.

msau42 reviewed (commented):
You'll also have to change `syncStates` to handle reconstructing block volumes.

mtanino commented on pkg/volume/util/util.go:
```diff
@@ -235,3 +235,49 @@ func LoadPodFromFile(filePath string) (*v1.Pod, error) {
 	return pod, nil
 }
+
+func GetVolumeModeForVolume(pod *v1.Pod, kubeClient clientset.Interface, pv *v1.PersistentVolume) (v1.PersistentVolumeMode, error) {
```
hmm, seems dead method. Will remove.

mtanino commented on pkg/volume/volume.go:
```diff
@@ -132,6 +132,30 @@ type Unmounter interface {
+	//
+	GetVolumePath() string
+	//
+	GetVolumeMode() v1.PersistentVolumeMode
```
Originally I couldn't find the way to access Persistent Volume Claim at makeBlockVolumes() method. Therefore, I implemented this method to determine volumeMode.
But I suppose PV can be accessed from the makeBlockVolumes(), and now we can decide volumeMode from only PV object.
Therefore, I think we don't need this method anymore.

Correct?

mtanino replied:
never mind. Since we added volumeDevices struct, we don't need to check volumeMode when we add volume to device list.

msau42 replied:
But somewhere (probably volume manager and not the plugin), we do need check for the error conditions: volumeDevices pvc.volumeMode = filesystem, and volumeMounts pvc.volumeMode = block.

mtanino replied:
@msau42 
Probably it's better to check these parameter combination before attaching volume to kubelet node.
IIUC, can we check these combinations in the WaitForAttachAndMount() of volume_manager.go?

```
func (vm *volumeManager) WaitForAttachAndMount(pod *v1.Pod) error {
```

msau42 replied:
Agree, it should be checked before invoking mount or map.

mtanino commented on pkg/kubelet/volumemanager/reconciler/reconciler.go:
```diff
@@ -268,7 +303,6 @@ func (rc *reconciler) reconcile() {
 		if !rc.desiredStateOfWorld.VolumeExists(attachedVolume.VolumeName) &&
 			!rc.operationExecutor.IsOperationPending(attachedVolume.VolumeName, nestedpendingoperations.EmptyUniquePodName) {
 			if attachedVolume.GloballyMounted {
-				// Volume is globally mounted to device, unmount it
```
removed unexpectedly. will add again.

mtanino commented on pkg/kubelet/volumemanager/reconciler/reconciler.go:
```diff
@@ -239,12 +252,34 @@ func (rc *reconciler) reconcile() {
+				glog.Infof("#### DEBUG LOG ####: reconcile attached/mounted, MountVolume pv: %v", pv)
+				glog.Infof("#### DEBUG LOG ####: reconcile attached/mounted, MountVolume pvcVolumeMode: %s", pvcVolumeMode)
+			} else {
+				glog.V(12).Infof(volumeToMount.GenerateMsgDetailed("Starting operationExecutor.MountVolume", remountingLogStr))
```
ack.

mtanino commented on pkg/kubelet/volumemanager/reconciler/reconciler.go:
```diff
@@ -239,12 +252,34 @@ func (rc *reconciler) reconcile() {
+			var pvcVolumeMode v1.PersistentVolumeMode
+			pv := volumeToMount.VolumeSpec.PersistentVolume
+			if pv != nil {
+				pvcVolumeMode, err = volumeutil.GetVolumeModeFromPVC(volumeToMount.Pod.ObjectMeta.Namespace, rc.kubeClient, pv)
```
ack.

msau42 commented:
/assign

jsafrane commented on pkg/kubelet/container/runtime.go:
```diff
@@ -449,6 +449,8 @@ type RunContainerOptions struct {
 type VolumeInfo struct {
 	// Mounter is the volume's mounter
 	Mounter volume.Mounter
+	// BlockVolumeMapper is the Block volume's mounter
```
mounter -> mapper

jsafrane commented on pkg/kubelet/volumemanager/volume_manager.go:
```diff
@@ -334,6 +335,31 @@ func (vm *volumeManager) WaitForAttachAndMount(pod *v1.Pod) error {
 		return nil
 	}
 
+	volumeModeMap, err := makeVolumeModeMap(pod, vm.kubeClient)
```
This new code checks that pod's VolumeDevices refers to Block PVCs. It's correct, but why it's in `WaitForAttachAndMount`? And should it check also `VolumeMounts` too?

IMO this check should be somewhere else. There are admission hooks in kubelet that allow a pod to run on the node or they refuse to run the pod with an error message to the user. Maybe a new handler that would check modes of all PVCs and VolumeMounts and VolumeDevices would be more appropriate. Someone from sig-node should confirm that PVCs are bound at this point (IMO they are...)

mtanino replied:
@jsafrane 
Thank you for the suggestion. Let me investigate the admission hooks.
The purpose of this check is invalid parameter combinations should be return as error before as early as possible.(At least, we should return error before creating a volume)
If we can check this at API validation phase, it's nice but seems we can't check this at validation phase.

>> And should it check also VolumeMounts too?

Agree.

msau42 replied:
In the proposal PR, the discussion concluded that we can't verify at API validation or admission controller.  So we have to do it early in kubelet.

mtanino replied:
@msau42 thanks, so we should check this in the WaitForAttachAndMount(), right?

msau42 replied:
So the check for if PVC is bound is done in [here](https://github.com/kubernetes/kubernetes/blob/50ec4384ff0ac482beb164767e562765ad82bde7/pkg/kubelet/volumemanager/populator/desired_state_of_world_populator.go#L339).  We don't add the volume to dsw PVC is not bound.

We could potentially extend this check to verify the `volumeMode`.  I would also generate an event on the Pod with the error.  Otherwise they have to wait for the attach/mount failure with a generic "timed out" message.

jsafrane replied:
Kubelet has its own admission, it's something else than admission controller on the API server.  Kubelet admission runs after a pod gets scheduled on a node and kubelet can say "no, I won't run this" from whatever reason. The pod is then Pending forever with appropriate events / messages. 

Look for `PodAdmitHandler` interface in kubelet and all `AddPodAdmitHandler` calls out there.

Note that I just know that this mechanism exists, I don't know if it's the right tool to use here. Ask someone from sig-node.

msau42 replied:
Pod rejection will send it back to the scheduler and will cause it to endlessly loop between scheduler and kubelet.

jsafrane commented on pkg/kubelet/volumemanager/reconciler/reconciler.go:
```diff
@@ -171,10 +172,22 @@ func (rc *reconciler) reconcile() {
+				glog.Infof("#### DEBUG LOG ####: reconcile mountedVolume.Mounter case")
+				//glog.Infof("#### DEBUG LOG ####: reconcile mountedVolume.Mounter: %s", mountedVolume.Mounter)
+			} else if mountedVolume.BlockVolumeMapper != nil {
+				// Volume is moapped, unmap it
```
moapped -> mapped

jsafrane commented on pkg/kubelet/volumemanager/reconciler/reconciler.go:
```diff
@@ -268,10 +300,23 @@ func (rc *reconciler) reconcile() {
 		if !rc.desiredStateOfWorld.VolumeExists(attachedVolume.VolumeName) &&
 			!rc.operationExecutor.IsOperationPending(attachedVolume.VolumeName, nestedpendingoperations.EmptyUniquePodName) {
 			if attachedVolume.GloballyMounted {
```
who fills `attachedVolume.GloballyMounted` for block devices? IMO it's filled at the end of `attacher.MountDevice`, but that applies only to filesystem volumes, right?

jsafrane replied:
found it, `GenerateMapVolumeFunc`

mtanino replied:
The attachedVolume.GloballyMounted is set via MarkDeviceAsMounted().
For filesystem volume case, MarkDeviceAsMounted() is called in the GenerateMountVolumeFunc().
And for Block volume case,  MarkDeviceAsMounted() is called in the GenerateMapVolumeFunc().

```
- GenerateMountVolumeFunc()
    -> volumeAttacher.MountDevice()
    -> MarkDeviceAsMounted()

- GenerateMapVolumeFunc()
     -> mapDevice()
     -> MarkDeviceAsMounted()
```

Commit:
BlockVolumesSupport: CRI, VolumeManager and OperationExecutor changes

This patch contains following changes.
- container runtime changes for adding block devices
- volumemanager changes
- operationexecutor changes

mtanino commented:
Reference count scheme for symbolic link per pod is added.

TODOs:
- Change syncStates()
- Check for volumeDevices, volumeMounts and volumeMode combinations
- Add comments and remove debug code
- Add more error handling code
- Add test code, test changes

mtanino commented:
`Done:`

- Add symbolic link refcount logic, Move out symbolic link creation and deletion logic from plugin
- Update syncStates() and reconstruct() logic for block volume
- Remove pod parameter from attachedVolume

`TODO:`
- volumeDevices{} and PVC combination check(still under consideration the place to implement it)
- Add comments and remove debug code
- Add more error handling code
- Add test code, test changes

msau42 commented on pkg/volume/util/util.go:
```diff
@@ -270,3 +275,148 @@ func stringToSet(str, delimiter string) (sets.String, error) {
+	}
+
+	// Check symbolic link is exist
+	fi, err := os.Lstat(mapPath + "/" + string(linkName))
```
Make this a variable since you use it a few times

msau42 commented on pkg/volume/util/util.go:
```diff
@@ -270,3 +275,148 @@ func stringToSet(str, delimiter string) (sets.String, error) {
+	glog.Errorf("#### DEBUG LOG ####: unmapDevice mapPath; %s", mapPath)
+
+	// Check map path is exist
+	if pathExists, pathErr := PathExists(mapPath); pathErr != nil {
```
Do you need this check?  Will it be covered by Lstat below?

msau42 commented on pkg/volume/util/util.go:
```diff
@@ -270,3 +275,148 @@ func stringToSet(str, delimiter string) (sets.String, error) {
+		return err
+	}
+	glog.Errorf("#### DEBUG LOG ####: unmapDevice find symlink")
+	if fi.Mode()&os.ModeSymlink != os.ModeSymlink {
```
add spacing around the &

msau42 commented on pkg/volume/util/util.go:
```diff
@@ -270,3 +275,148 @@ func stringToSet(str, delimiter string) (sets.String, error) {
 }
+
+// MapDevice creates a symbolic link to block device under specified map path
+func MapDevice(devicePath string, mapPath string, linkName string) error {
```
Would mapDir be more clear?  When I see path, I often assume full path.

msau42 commented on pkg/volume/util/util.go:
```diff
@@ -270,3 +275,148 @@ func stringToSet(str, delimiter string) (sets.String, error) {
+		return fmt.Errorf("Failed to remove global map path directory. mapPath is empty")
+	}
+	glog.Errorf("#### DEBUG LOG ####: RemoveGlobalMapPath mapPath; %s", mapPath)
+	err := os.RemoveAll(mapPath)
```
When would you still have symlinks remaining inside?

msau42 commented on pkg/volume/util/util.go:
```diff
@@ -270,3 +275,148 @@ func stringToSet(str, delimiter string) (sets.String, error) {
+}
+
+// GetDeviceSymlinkRefs searches symbolic links under global map path
+func GetDeviceSymlinkRefs(devPath string, globalVDPDPath string) ([]string, error) {
```
what does VDPD stand for?

mtanino replied:
This means Global volume device plugin dir. I'll update the comment.

msau42 commented on pkg/volume/util/util.go:
```diff
@@ -270,3 +275,148 @@ func stringToSet(str, delimiter string) (sets.String, error) {
+	// Check symbolic link is exist
+	fi, err := os.Lstat(mapPath + "/" + string(linkName))
+	if err != nil && os.IsNotExist(err) {
+		return newSymlinkNotContained(mapPath + "/" + string(linkName))
```
How about NotExist instead?

msau42 commented on pkg/volume/util/util.go:
```diff
@@ -270,3 +275,148 @@ func stringToSet(str, delimiter string) (sets.String, error) {
+	}
+	glog.Errorf("#### DEBUG LOG ####: unmapDevice find symlink")
+	if fi.Mode()&os.ModeSymlink != os.ModeSymlink {
+		return newSymlinkNotContained(mapPath + "/" + string(linkName))
```
Do you want to distinguish between not exist and not a symlink?

msau42 commented on pkg/volume/volume.go:
```diff
@@ -132,6 +132,48 @@ type Unmounter interface {
+	GetDevicePath() (string, error)
+	// GetGlobalMapPath generates a global map path to create
+	// a symbolic link to a block device.
+	GetGlobalMapPath(spec *Spec) (string, error)
```
Why does the plugin need to implement these two methods? Shouldn't volume manager be able to generate the paths without plugin help?

mtanino replied:
Global map path contains plugin own unique directory name like below.
`28000001ff0414e2-lun-0 => WWN + lun number` 
This is similar to current filesystem case.
Therefore, volume manager can't generate this path by itself. If we use another plugin independent unique volume information for this path(ex volume name?), plugin doesn't need to implement this method.

```
# find plugins | ag devices 
plugins/kubernetes.io/fc/volumeDevices
plugins/kubernetes.io/fc/volumeDevices/28000001ff0414e2-lun-0
plugins/kubernetes.io/fc/volumeDevices/28000001ff0414e2-lun-0/4c65e761-a264-11e7-bf52-525400f59db4
```

GetPodDeviceMapPath() returns similar value to GetPath() in each plugin.
Volume manager might generate this value using volume spec. 

```
func (fc *fcDisk) GetPath() string {
        name := fcPluginName
        // safe to use PodVolumeDir now: volume teardown occurs before pod is cleaned up
        return fc.plugin.host.GetPodVolumeDir(fc.podUID, utilstrings.EscapeQualifiedNameForDisk(name), fc.volName)
}
```

Pod device map path
```
# find pods | ag devices
pods/4c65e761-a264-11e7-bf52-525400f59db4/volumeDevices
pods/4c65e761-a264-11e7-bf52-525400f59db4/volumeDevices/kubernetes.io~fc
pods/4c65e761-a264-11e7-bf52-525400f59db4/volumeDevices/kubernetes.io~fc/pv0001
pods/4c65e761-a264-11e7-bf52-525400f59db4/volumeDevices/kubernetes.io~fc/pv0001/symlink
```

msau42 replied:
Ah thanks for the examples, it helps a lot.  Can you add comments with the format of the various paths?

msau42 commented on pkg/volume/util/operationexecutor/operation_generator.go:
```diff
@@ -627,6 +627,289 @@ func (og *operationGenerator) GenerateUnmountDeviceFunc(
+		}
+		// Device mapping for global map path succeeded
+		simpleMsg, detailedMsg := volumeToMount.GenerateMsg("MapVolume.MapDevice succeeded", fmt.Sprintf("globalMapPath %q", globalMapPath))
+		verbosity := glog.Level(1)
```
For debugging?

msau42 commented on pkg/volume/util/operationexecutor/operation_generator.go:
```diff
@@ -627,6 +627,289 @@ func (og *operationGenerator) GenerateUnmountDeviceFunc(
+
+		// Map device to pod device map path under the given pod directory using symbolic link
+		volumeMapPath := blockVolumeMapper.GetPodDeviceMapPath()
+		mapErr = util.MapDevice(devicePath, volumeMapPath, util.PodVolumeDirSymlinkName)
```
Should the symlink name be the volume name instead of a const string?  You can have multiple volumes in the same Pod.

msau42 replied:
Can pv0001 itself be a symlink?

msau42 commented on pkg/volume/util/operationexecutor/operation_executor.go:
```diff
@@ -105,6 +105,26 @@ type OperationExecutor interface {
+	// MapVolume is used when the volumeMode is 'Block'. This method creates a symbolic link to the block device on global map path.
+	// Specifically it will:
+	// * Wait for the device to finish attaching (for attachable volumes only).
+	// * Map volume to global map path using symbolic link (for attachable volumes only).
```
This is for both attachable and non-attachable right?

msau42 commented on pkg/volume/util/operationexecutor/operation_generator.go:
```diff
@@ -627,6 +627,289 @@ func (og *operationGenerator) GenerateUnmountDeviceFunc(
+			// On failure, return error. Caller will log and retry.
+			return volumeToUnmount.GenerateErrorDetailed("UnmapVolume.UnmapDevice on pod device map path failed", unmapDeviceErr)
+		}
+		removeMapPathErr := util.RemoveMapPath(podDeviceUnmapPath)
```
What if you have multiple volumes under the same directory?

msau42 commented on pkg/volume/util/operationexecutor/operation_generator.go:
```diff
@@ -627,6 +627,289 @@ func (og *operationGenerator) GenerateUnmountDeviceFunc(
+
+	return func() error {
+		// Execute unmap
+		unmapErr := blockVolumeUnmapper.TearDownDevice()
```
Should the TearDown happen after you delete the symlinks?

msau42 commented on pkg/volume/util/operationexecutor/operation_executor.go:
```diff
@@ -105,6 +105,26 @@ type OperationExecutor interface {
+	UnmapVolume(volumeToUnmount MountedVolume, actualStateOfWorld ActualStateOfWorldMounterUpdater) error
+
+	// UnmapDevice unmaps the volumes global map path from the device (for
+	// attachable volumes only, freeing it for detach. It then updates the
```
Both attachable and non-attachable right?

msau42 commented on pkg/volume/util/operationexecutor/operation_executor.go:
```diff
@@ -105,6 +105,26 @@ type OperationExecutor interface {
+	// volumeToUnmount and global map path. And then, updates the actual state of the world to reflect that.
+	UnmapVolume(volumeToUnmount MountedVolume, actualStateOfWorld ActualStateOfWorldMounterUpdater) error
+
+	// UnmapDevice unmaps the volumes global map path from the device (for
```
IIUC, it doesn't actually unmap anything, it just checks if all references to the device have been removed?

msau42 commented on pkg/volume/volume.go:
```diff
@@ -132,6 +132,48 @@ type Unmounter interface {
+	TearDownDeviceAt(dir string) error
+	// GetGlobalUnmapPath deletes a symbolic link to a block device
+	// under a global map path
+	GetGlobalUnmapPath(spec *Spec) (string, error)
```
Same thing, can volume manager generate these 2 paths?

msau42 commented on pkg/kubelet/kubelet_pods.go:
```diff
@@ -111,6 +111,24 @@ func (kl *Kubelet) makeDevices(pod *v1.Pod, container *v1.Container) ([]kubecont
+		if !ok || vol.BlockVolumeMapper == nil {
+			continue
+		}
+		devicePath, err := vol.BlockVolumeMapper.GetDevicePath()
```
Should the devicePath be the symlink under the Pod's directory instead?

mtanino commented:
@msau42 @jsafrane @erinboyd @screeley44
All comments from @msau42 are addressed. PTAL.

TODO:
- Remove debug code
- Add test code, test changes

msau42 commented on pkg/kubelet/kubelet_pods.go:
```diff
@@ -111,6 +112,30 @@ func (kl *Kubelet) makeDevices(pod *v1.Pod, container *v1.Container) ([]kubecont
+		vol, ok := podVolumes[device.Name]
+		if !ok || vol.BlockVolumeMapper == nil {
+			glog.Warningf("Block volume cannot be satisfied for container %q, because the volume is missing or the volume mapper is nil: %q", container.Name, device)
+			continue
```
I'm a bit surprised we do this for filesystem volumes too.  Wouldn't it be a problem if we started a container with not all the volumes mounted, and the application wrote to the path expecting the data to be on a persisted volume?  I guess we don't create a path for it, so the application would get a PathNotFound error first.  But it's possible the user could mount the volume at an existing path too, in which case this would be an issue.

mtanino replied:
I may be missing something?
But since this method is iterating VolumeDevices, there isn't a case that the pod volume is filesystem volume.

msau42 replied:
Opened #53421 to handle the fs case

msau42 commented on pkg/kubelet/volumemanager/populator/desired_state_of_world_populator.go:
```diff
@@ -382,6 +383,26 @@ func (dswp *desiredStateOfWorldPopulator) createVolumeSpec(
 			pvcUID)
 
+		// Error if a container doesn't have volumeDevices but the volumeMode of PVC is Block
+		if len(devicesMap) == 0 && volumeMode == v1.PersistentVolumeBlock {
```
Can this check be merged with the 2nd check?  You just need to check if there's an entry in the map && mode is block

mtanino replied:
will do.

msau42 commented on pkg/kubelet/volumemanager/populator/desired_state_of_world_populator.go:
```diff
@@ -419,15 +441,15 @@ func (dswp *desiredStateOfWorldPopulator) getPVCExtractPV(
 	}
 
-	return pvc.Spec.VolumeName, pvc.UID, nil
+	return pvc.Spec.VolumeName, pvc.UID, *pvc.Spec.VolumeMode, nil
```
Can VolumeMode be nil?  I think @screeley44 is writing a helper function to get the volume mode that will default it.

mtanino replied:
yes. I also remember that I have similar method, but I'll use @screeley44 's one.

msau42 commented on pkg/kubelet/volumemanager/populator/desired_state_of_world_populator.go:
```diff
@@ -468,3 +490,19 @@ func getPVVolumeGidAnnotationValue(pv *v1.PersistentVolume) string {
+		}
+		for _, device := range container.VolumeDevices {
+			glog.Infof("#### DEBUG LOG ####: makeVolumeDevicesMap: device: %v", device)
+			volumeDevicesMap[device.Name] = true
```
Is it possible to have the same device name in two different containers, but referencing two different volumes?

msau42 commented on pkg/kubelet/volumemanager/reconciler/reconciler.go:
```diff
@@ -574,31 +651,41 @@ func getVolumesFromPodDir(podDir string) ([]podVolume, error) {
+		//     block volume:      /pods/{podUid}}/volumeDevices/{escapeQualifiedPluginName}/{volumeName}
+		volumesDirs := []string{path.Join(podDir, options.DefaultKubeletVolumesDirName),
+			path.Join(podDir, options.DefaultKubeletVolumeDevicesDirName)}
+		for _, volumesDir := range volumesDirs {
```
can volumesDir be a map with key = volumeMode, and value = path?

mtanino replied:
done.

msau42 commented on pkg/volume/util/util.go:
```diff
@@ -270,3 +270,119 @@ func stringToSet(str, delimiter string) (sets.String, error) {
+	if err = os.MkdirAll(mapDir, 0750); err != nil {
+		return fmt.Errorf("Failed to mkdir %s, error", mapDir)
+	}
+	// If old symlink exits, remove it then create new one
```
If it exists, does it mean we didn't clean up correctly?

mtanino replied:
@msau42 
I suppose that there is a case a node accidentally rebooted, and then kubelet recreates symlinks.

If the node is rebooted, order of /dev/sdX may be changed, therefore we should remove old symlink, then create new symlink associated to a new /dev/sdX.

msau42 replied:
How does volume reconstruction fit in here?

msau42 commented on pkg/volume/volume.go:
```diff
@@ -132,6 +132,55 @@ type Unmounter interface {
+	// which may or may not exist yet and returns combination of physical
+	// device path of a block volume and error.
+	// This may be called more than once, so implementations must be idempotent.
+	SetUpDeviceAt(dir string) (string, error)
```
Do we need this?

mtanino replied:
done.

msau42 commented on pkg/volume/volume.go:
```diff
@@ -132,6 +132,55 @@ type Unmounter interface {
+	TearDownDevice() error
+	// TearDownDeviceAt removes traces of the SetUpDevice procedure under
+	// the specified directory.
+	TearDownDeviceAt(dir string) error
```
Do we need this?

mtanino replied:
done

msau42 commented on pkg/kubelet/volumemanager/reconciler/reconciler.go:
```diff
@@ -171,10 +172,21 @@ func (rc *reconciler) reconcile() {
-				mountedVolume.MountedVolume, rc.actualStateOfWorld)
+			var err error
+			pvVolumeMode := volumehelper.GetVolumeMode(mountedVolume.MountedVolume.VolumeSpec)
+			if pvVolumeMode != v1.PersistentVolumeBlock {
```
Instead of if/else everywhere, can we make an interface and have a fs implementation and a block implementation?

jsafrane commented on pkg/volume/util/util.go:
```diff
@@ -270,3 +270,119 @@ func stringToSet(str, delimiter string) (sets.String, error) {
+		if err != nil {
+			return nil, fmt.Errorf("Symbolic link cannot be retrieved %v", err)
+		}
+		if strings.Contains(path, devPath) {
```
Should be the strings equal?

What if the symlink points to another device? It's used to scan global directory of a volume and the volume should have only one device. As I understand it, something went terribly wrong if there are mixed symlinks to different devices in one global dir. Should we remove the stale ones?

mtanino replied:
So the big difference between mount and symlink, mount point is always equal to corresponding block device, but symbolic link might refer wrong block device.

(ex, after rebooting node, order of device might be changed. Kubelet recreate  symlink during start up, therefore I suppose this should not happen.)

Do you think we should not obtain physical device from destination of symbolic link?
If so, even if destination of symbolic link is wrong/corrupted, we don't detach unexpected volume. Then symlink is used for just counting number.

>> Should be the strings equal?
Agree. To avoid unexpected selecting a volume, this should check that the strings are equal.

Here is example of Global map path. There are multiple symlinks but destination of them are same.
```
Global map path for "Block" volumeMode volume
/var/lib/kubelet/plugins/kubernetes.io/{pluginName}/volumeDevices/{volumePluginDependentPath}/{pod uuid1}
/var/lib/kubelet/plugins/kubernetes.io/{pluginName}/volumeDevices/{volumePluginDependentPath}/{pod uuid2}
...
```

>> Should we remove the stale ones?

We may need this logic at somewhere. We can check existence of  pod using UUID.

jsafrane replied:
Can we require (or strongly recommend) that WaitForAttach() and SetupUpDevice() returns device name that's stable across reboots? For most devices it would be something in `/dev/disk/by-id`, I am not sure about all FC/iSCSI/multipath devices though.

> Do you think we should not obtain physical device from destination of symbolic link?

I don't know. What happens after reboot? Will kubelet run `SetUpDevice()` again and re-create the symlinks or will it assume that the device is already set up and it will just re-use the links (even though they may point to an old device)?

> If so, even if destination of symbolic link is wrong/corrupted, we don't detach unexpected volume. Then symlink is used for just counting number.

Yeah, links for reference counting is OK.

msau42 replied:
Requiring would be nice if we can enforce it (I don't see how).  Otherwise we have to assume the plugin could mess up.  And it's not just reboots, also adding/removing devices while the node is still running can cause issues if you don't use a stable path.  Perhaps, one volume at /dev/sdb could be removed, and a different volume (even different volume type) could be added at /dev/sdb afterwards.  Can we detect and protect against this?

On a kubelet restart, there's a few scenarios we have to consider when detecting existing links and whether or not we can trust what it's pointing to:
1. Pod is still running on the node: The pod will be added to dsw, and in the reconciler, it will go through setup, then update asw.  This should be safe because we have the full volume spec and can override stale links.
2. Pod is deleted: Volume reconstruction scans pod directories for volume links and adds the volume to asw.  Can we trust the links here?  Only if volume plugin implements reconstruction correctly?  What is the danger if the links are stale?

mtanino replied:
I'd add getDevicePath() into BlockVolumeMapper interface to obtain stable path.
We require that the getDevicePath() return path of symlink under /dev/disk/by-id/, and the check return value if it's contains /dev/disk/by-id path.

>>Pod is still running on the node: The pod will be added to dsw, and in the reconciler, it will go through setup, then >>update asw. This should be safe because we have the full volume spec and can override stale links.

Agree.

>>Pod is deleted: Volume reconstruction scans pod directories for volume links and adds the volume to asw. Can we >>trust the links here? Only if volume plugin implements reconstruction correctly? What is the danger if the links are stale?

I think we shouldn't trust the destination of link. Should we store full name and path of /dev/disk/by-id/{link} on anywhere? If we have this, we can refer correct volume without volume spec.

>>also adding/removing devices while the node is still running can cause issues if you don't use a stable path. 
>>Perhaps, one volume at /dev/sdb could be removed, and a different volume (even different volume type) could be >> added at /dev/sdb afterwards. Can we detect and protect against this?

I confirmed that when we use block device at docker, we can't detect even if the attached device is silently replaced.
```
% sudo docker run --device=/dev/loop2:/dev/xvdc --rm -it centos bash
[root@9ad607444fe9 /]# md5sum /dev/xvdc
53c0a6941d5f08024fda66e2ae3c934d  /dev/xvdc

- After detach and replace backend file of /dev/loop2 from fcfile to fcfile2. Only checksum is different.

[root@9ad607444fe9 /]# md5sum /dev/xvdc
01732c77d962867ce7bd125830634a53  /dev/xvdc

- If device is gone, just fail to writing operation with No space left on device.
[root@9ad607444fe9 /]# dd if=/dev/zero of=/dev/xvdc
dd: writing to '/dev/xvdc': No space left on device
1+0 records in
0+0 records out
0 bytes (0 B) copied, 0.0178061 s, 0.0 kB/s
```

msau42 replied:
The Setup() already should return the stable path.  Where else do we need to get the stable path?

I don't think we can enforce "/dev/disk/by-id" path.  On some OSes, we may not have permission to modify it.

msau42 replied:
@mtanino can you also try this sequence?

1. Pass in unique symlink to docker, ie /dev/unique1 -> /dev/loop2
2. Remove the device
3. Add a new device with a different unique symlink, but points to the same device path, ie /dev/unique2 -> /dev/loop2

Is the container able to access the new device?

mtanino replied:
@msau42 
I taleked @msau42 directly, even if we follow above steps, container had access to newly added /dev/loop2. Therefore we guess container itself has resolved device path inside even if we pass symlink at first.

mtanino replied:
@msau42 @jsafrane

As for the volume reconstruction case 2(pod is deleted), I think we can believe newly created volumeSpec during ConstructVolumeSpec at each plugin.
I checked several volume plugins, they create new volume spec from only information of mountpath. Even if we introduce block volume support, I think we can use same way because newly introduced `global map path` still contain volume plugins unique information like this.

- ./kubernetes.io/fc/volumeDevices/`28000001ff0414e2-lun-0`/8576273a-ac4f-11e7-bc66-525400f59db4

But I think we need separated volume construction method like ConstructVolumeSpecForBlock() since path is different from filesystem case.

plugin | way to create new volume spec
-- | --
aws | get VolumeID from mountpoint path on plugin directory and create new volumespec from it
azure_dd | get DataDiskURI from mountpoint path on plugin directory
gce_pd | get PDName from mountpoint path on plugin directory
portworxglusterstorage os | create a new volumespec from volumename only(maybe volumename is unique key?)

jsafrane commented on pkg/volume/volume.go:
```diff
@@ -132,6 +132,55 @@ type Unmounter interface {
+	// If the plugin is attachable, the volume has been attached already,
+	// just return empty string for device path.
+	// This may be called more than once, so implementations must be idempotent.
+	SetUpDevice() (string, error)
```
From the description it's not clear to me what this function should do. If the plugin is attachable, it should not do anything, right? If the plugin is not attachable, it should prepare the device in /dev (or where appropriate) and return it's name.

mtanino replied:
correct. I'll update the comment.

jsafrane commented on pkg/volume/util/operationexecutor/operation_generator.go:
```diff
@@ -627,6 +627,286 @@ func (og *operationGenerator) GenerateUnmountDeviceFunc(
+
+	return func() error {
+		// Execute tear down device
+		unmapErr := blockVolumeUnmapper.TearDownDevice()
```
IMO this is too early to call TearDownDevice. Kubelet should make sure the device is not needed any longer (by counting symlinks) and tear it down only after that. Should it be part of GenerateUnmapDeviceFunc?

mtanino replied:
As for the filesystem volume case, TearDown() is called at the same timing.

https://github.com/kubernetes/kubernetes/pull/51494/files#diff-450e811a4953f760ff1594ede8b2037eL521

I think TearDown() expects that the mount point is unmounted, but I saw some plugin are doing detach operation in the TearDown, because TearDown() is a only method to cleanup attached volume if the plugin is non-attachable.

But why doesn't this cause problem currently?

Yes, I think we can move this around L873 in GenerateUnmapDeviceFunc()
https://github.com/kubernetes/kubernetes/pull/51494/files#diff-450e811a4953f760ff1594ede8b2037eR873

jsafrane replied:
To be honest, TearDown is a mystery to me too. @rootfs might educate us how it's supposed to work.

msau42 replied:
From what I've seen, the plugins that detach in the TearDown do their own refcounting, so only the last reference will detach.

But for block, the semantics for setup and teardown are different because the volume manager is the one setting up the Pod volume and global symlinks, not the plugin.

mtanino replied:
>> IMO this is too early to call TearDownDevice. Kubelet should make sure the device is not >>needed any longer (by counting symlinks) and tear it down only after that. Should it be part of GenerateUnmapDeviceFunc?

@jsafrane @msau42 
Thank you for the explanation. Now I think the TearDownDevice() should be done after deviceOpened check is passed in GenerateUnmapDeviceFunc().
Is this correct?

jsafrane reviewed (commented):
I did some review, but I got lost in the a/d operations... Take with a grain of salt.

mtanino commented on pkg/kubelet/kubelet_pods.go:
```diff
@@ -111,6 +112,30 @@ func (kl *Kubelet) makeDevices(pod *v1.Pod, container *v1.Container) ([]kubecont
+		}
+		// Get a symbolic link associated to a block device under pod device path
+		dirPath, volName := vol.BlockVolumeMapper.GetPodDeviceMapPath()
+		symlinkPath := dirPath + "/" + volName
```
done.

mtanino commented on pkg/kubelet/kubelet_pods.go:
```diff
@@ -400,12 +425,19 @@ func (kl *Kubelet) GenerateRunContainerOptions(pod *v1.Pod, container *v1.Contai
 	// TODO(random-liu): Move following convert functions into pkg/kubelet/container
-	devices, err := kl.makeDevices(pod, container)
+	devices, err := kl.makeGPUDevices(pod, container)
+
```
done.

mtanino commented:
TODOs

- Ensure to use correct block device using symbolic link
- Consider volume reconstruct case
- Consider proper place to call TearDownDevice()
- Use  @screeley44's helper function to get volumeMode
- Address this comment "can volumesDir be a map with key = volumeMode, and value = path?
" in reconciler
- Remove debug code
- Add test code, test changes

mtanino commented:
Done:
- Address this comment "can volumesDir be a map with key = volumeMode, and value = path?
" in reconciler

TODOs

- Ensure to use correct block device using symbolic link
- Consider volume reconstruct case
- Consider proper place to call TearDownDevice()
- Use @screeley44's helper function to get volumeMode
- Remove debug code
- Add test code, test changes

mtanino commented:
Done:
- To prevent silent volume replacement using loopback
- Add constructBlockVolume() for block specific volume reconstruction and update blockvolumeplugin interface

TODOs
- Move TearDownDevice() into GenerateUnmapDeviceFunc()
- Use @screeley44's helper function to get volumeMode
- Add test code, test changes

mtanino commented:
Done:
- Move TearDownDevice() into GenerateUnmapDeviceFunc()
- Use @screeley44's helper function to get volumeMode

TODOs
- Add test code, test changes

mtanino commented:
@msau42 @jsafrane @erinboyd @screeley44

I think I have addressed all feedback. PTAL.
Thanks.

mtanino commented:
Done:
- Introduce BlockVolume interface

TODOs
- Add test code, test changes

mtanino commented on pkg/volume/util/operationexecutor/operation_generator.go:
```diff
@@ -56,19 +56,23 @@ type operationGenerator struct {
 	// which verifies that the components (binaries, etc.) required to mount
 	// the volume are available on the underlying node before attempting mount.
 	checkNodeCapabilitiesBeforeMount bool
+	//
```
Comment will be added.

jingxu97 replied:
ping here

mtanino commented:
Updates based on @jsafrane 's review.

- Change first argument of ConstructBlockVolumeSpec from podName to podUID
- Add arguments mapPath and devicePath to TearDownDevice()
- Remove to pass exec from loopdevice operation methods
- Update IsSymlinkExist()
- Remove unnecessary symlink check before removing file
- Test: add volumetesting.VerifyGetPodDeviceMapPathCallCount check
- Fix typos

mtanino commented:
@msau42 @jsafrane 
This PR is ready to review. PTAL

msau42 commented on pkg/kubelet/volumemanager/reconciler/reconciler_test.go:
```diff
@@ -399,6 +431,419 @@ func Test_Run_Positive_VolumeUnmountControllerAttachEnabled(t *testing.T) {
+	}
+
+	volumeSpec := &volume.Spec{
+		Volume:           &pod.Spec.Volumes[0],
```
Why do you need this?

msau42 commented on pkg/volume/util/operationexecutor/operation_generator.go:
```diff
@@ -627,6 +633,315 @@ func (og *operationGenerator) GenerateUnmountDeviceFunc(
+		}
+
+		// Device mapping for pod device map path succeeded
+		simpleMsg, detailedMsg = volumeToMount.GenerateMsg("MapVolume.MapDevice succeeded", fmt.Sprintf("volumeMapPath %q", volumeMapPath))
```
This was already logged above? Which one do you want to keep?

msau42 commented on pkg/volume/util/operationexecutor/operation_generator.go:
```diff
@@ -627,6 +633,315 @@ func (og *operationGenerator) GenerateUnmountDeviceFunc(
+		}
+
+		glog.Infof(
+			"UnmapVolume.TearDown succeeded for volume %q (OuterVolumeSpecName: %q) pod %q (UID: %q). InnerVolumeSpecName %q. PluginName %q, VolumeGidValue %q",
```
msg is not correct

msau42 commented on pkg/kubelet/volumemanager/reconciler/reconciler.go:
```diff
@@ -437,6 +429,31 @@ func (rc *reconciler) reconstructVolume(volume podVolume) (*reconstructedVolume,
+	}
+
+	var volumeSpec *volumepkg.Spec
+	if volume.volumeMode == v1.PersistentVolumeBlock {
```
Can this section also go into the handler interface?

msau42 commented on pkg/volume/util/util.go:
```diff
@@ -270,3 +280,278 @@ func stringToSet(str, delimiter string) (sets.String, error) {
+}
+
+// GetLoopDevice returns the full path to the loop device associated with the given path.
+func (v VolumePathHandler) GetLoopDevice(path string) (string, error) {
```
Do the losetup utils need to go into a linux-specific file?

mtanino replied:
OK. good idea.

mtanino commented:
@msau42 @jsafrane
Updated based on @msau42's review. PTAL

msau42 commented on pkg/volume/util/operationexecutor/operation_generator.go:
```diff
@@ -627,6 +633,315 @@ func (og *operationGenerator) GenerateUnmountDeviceFunc(
+			return deviceToDetach.GenerateErrorDetailed("UnmapDevice failed", err)
+		}
+
+		// The block volume is not referenced from Pods. Release file descriptor lock.
```
Can the loopback device cleanup be done at the very end?  Right now, if we remove the loopback device first and subsequent steps fail, then the race condition can be exposed.

Also, what about the scenario where you have multiple maps and unmaps happening on the same volume at the same time?  Could a pod map a volume right before another thread removes the loopback device?  Maybe the loopback remove should not happen until after we mark the volume as unmounted.

msau42 replied:
Actually maybe this is ok.  Does the operation generator ensure that only one operation per volume can be run at a time?

mtanino replied:
@msau42 
I thought reconcier() is running on single thread.
Also during reconcier() loop, it seems that reconcier() ensure these set of operation generator's operations like (1) or (2) are executed as a series. As a result during (1), (2) can't interrupt(doesn't happen) or vice versa.

(1). pod create with vol
- make volume as attached
- make volume as mounted

(2). pod delete with vol
- make volume as unmounted
- make volume as detached (if other pods don't use this volume anymore)

msau42 replied:
Even if only one operation executes at a time, it is possible that it only partially completed, if it encountered an error.

mtanino replied:
@msau42 

>> it is possible that it only partially completed, if it encountered an error.

I agree this possibly happens. Could you let me know your supposed race scenario you mentioned above? I'd like to know what race can be prevented if we move loopback device cleanup to end of the method.

Here are rough steps of UnmapDevice steps.
```
GenerateUnmapDeviceFunc()

(1) check symlink references
  refs>0 -> return err, restart from step.1
  refs =0 -> go step.2

(2) check loopdevice existense
  yes -> remove loopdevice
  no -> warning message

(3) isDeviceOpened check
  still opened -> return err, restart from step.1
  not opened -> go step.4

(4) TearDownDevice
  success -> go step.5
  fail -> return err, restart from step.1

(5) RemoveMapPath(remove globalMapPath dir)
  success or notExist -> go step.6
  fail -> return err, restart from step.1

(6) MarkDeviceAsUnmounted
  success -> return nil(Complete GenerateUnmapDeviceFunc)
  fail -> return err, restart from step.1
```

msau42 replied:
I'm mainly concerned if you remove the loopback device before removing the global mount, would that open up the possibility of the device from being swapped out?

jingxu97 replied:
Looks like you haven't updated the code to match the operation ordering mentioned above? Could you also add some comments about the reason of this ordering?

msau42 commented on pkg/volume/util/operationexecutor/operation_generator.go:
```diff
@@ -627,6 +633,315 @@ func (og *operationGenerator) GenerateUnmountDeviceFunc(
+	}, blockVolumePlugin.GetPluginName(), nil
+}
+
+func (og *operationGenerator) GenerateUnmapDeviceFunc(
```
@jingxu97 can you help review this?  There's a lot cleanup steps going on here and we need to consider the cases where mounts could also be going on at the same time.

jingxu97 replied:
For each volume, if operation_executor call this function with a unique podName, then two UnmapDevice operations from different pods could be executed concurrently. If you want to make sure for a single volume, only one UnmapDevice operation is executed at one time, you could use EmptyUniquePodName. See https://github.com/kubernetes/kubernetes/blob/master/pkg/volume/util/operationexecutor/operation_executor.go#L677

msau42 commented on pkg/kubelet/kubelet_pods.go:
```diff
@@ -111,6 +112,36 @@ func (kl *Kubelet) makeDevices(pod *v1.Pod, container *v1.Container) ([]kubecont
 
+// makeBlockVolumes maps the raw block devices specified in the path of the container
+// Experimental
+func (kl *Kubelet) makeBlockVolumes(pod *v1.Pod, container *v1.Container, podVolumes kubecontainer.VolumeMap) ([]kubecontainer.DeviceInfo, error) {
```
Needs unit tests

mtanino replied:
ack. will do.

msau42 commented on pkg/kubelet/volumemanager/cache/actual_state_of_world_test.go:
```diff
@@ -318,6 +328,15 @@ func Test_AddPodToVolume_Negative_VolumeDoesntExist(t *testing.T) {
 			err)
 	}
+
+	blockplugin, err := volumePluginMgr.FindMapperPluginBySpec(volumeSpec)
```
Why does this test need to FindMapperPlugin but the others don't?

mtanino replied:
Since fake volume plugin has both NewMounter and NewBlockVolumeMapper methods, FindxxxBySpec doesn't need in other methods.

But I thought this method is trying to test FindPluginBySpec using volumeSpec.
Because volumetesting.GetTestVolumePluginMgr returns fake volume plugin, but line 301 is intentionally masking plugin from return. Therefore, I also added FindMapperPluginBySpec() to obtain plugin from volumeSpec.
Is this make sense?

msau42 replied:
Do we have test cases where block plugin is nil? since the majority of volume plugins are not going to implement the block interface yet.

mtanino replied:
I'll add test cases to check plugin nil case.

msau42 commented on pkg/kubelet/volumemanager/populator/desired_state_of_world_populator_test.go:
```diff
@@ -37,54 +39,144 @@ import (
+
+}
+
+func TestFindAndAddNewPods_FindAndRemoveDeletedPods_valid_block_volumeDevices(t *testing.T) {
```
Use Uppercase names

mtanino replied:
ack.

msau42 commented on pkg/kubelet/volumemanager/reconciler/reconciler.go:
```diff
@@ -171,10 +173,8 @@ func (rc *reconciler) reconcile() {
-			glog.V(12).Infof(mountedVolume.GenerateMsgDetailed("Starting operationExecutor.UnmountVolume", ""))
-			err := rc.operationExecutor.UnmountVolume(
-				mountedVolume.MountedVolume, rc.actualStateOfWorld)
+			mountedVolume.VolumeHandler = operationexecutor.NewVolumeHandler(mountedVolume.VolumeSpec, rc.operationExecutor)
```
If you're always creating the volume handler right before the operation, then there's no need to store it as a field in the volume struct.

mtanino replied:
agree. I confirmed that handler are always created right before the operation, should be removed from volume struct.

mtanino commented:
@msau42 
I've addressed to all your review comments. PTAL
```
Volumemanager updates  … e0881c4
- Add test case for makeBlockVolumes
- Add test cases for MapVolume, UnmapVolume, UnmapDevice
  to test plugin not found cases
- Remove VolumeStateHandler from volume struct
- Use upper case for test method name
- Fix typos

Volumemanager updates  … 7d25b39
- Add GetPersistentVolumeClaimVolumeMode() to volumehelper
```

msau42 commented on pkg/volume/util/volumehelper/volumehelper.go:
```diff
@@ -151,3 +151,13 @@ func GetVolumeMode(volumeSpec *volume.Spec) v1.PersistentVolumeMode {
+
+// GetPersistentVolumeClaimVolumeMode returns VolumeMode. If no VolumeMode was
+// requested, it returns "Filesystem" VolumeMode by default.
+func GetPersistentVolumeClaimVolumeMode(claim *v1.PersistentVolumeClaim) v1.PersistentVolumeMode {
```
We should not default any values now that the API is defaulted.

liggitt replied:
the defaulting is only applied if the feature gate is on, right? so this field could be nil

msau42 replied:
Yes there needs to be a nil check.  But if it's nil, we should not default a value.

mtanino commented on pkg/volume/util/operationexecutor/operation_executor.go:
```diff
@@ -105,6 +107,28 @@ type OperationExecutor interface {
+	// * Map volume to global map path using symbolic link.
+	// * Map the volume to the pod device map path using symbolic link.
+	// * Update actual state of world to reflect volume is mounted/mapped to the pod path.
+	MapVolume(waitForAttachTimeout time.Duration, volumeToMount VolumeToMount, actualStateOfWorld ActualStateOfWorldMounterUpdater) error
```
@jingxu97 @msau42 

Comments picking up from #54752.
```
jingxu97 15 hours ago  Member
do we need to make it public. Seems like all MountVolume or MapVolume in operation executor could be private?
```
I believe we could make all methods in OperationExecutor interfaceas as private since they always used via operation executor, but I think we shouldn't change all of them in this one PR.
Related methods are following. I agree on making them as private in this PR.

- MountVolume, MapVolume
- UnmountVolume, UnmapVolume
- UnmountDevice, UnmapDevice

mtanino commented on pkg/kubelet/volumemanager/populator/desired_state_of_world_populator_test.go:
```diff
@@ -156,6 +252,210 @@ func TestFindAndAddNewPods_FindAndRemoveDeletedPods(t *testing.T) {
+
+	// Assert
+	if volumeSpec == nil || err != nil {
+		t.Fatalf("Failed to create volumeSpec with combination of filesystem mode and volumeMounts. err: %v", err)
```
Comment is wrong.
`t.Fatalf("Failed to create volumeSpec with combination of block mode and volumeDevices. err: %v", err)`

msau42 commented on pkg/kubelet/volumemanager/reconciler/reconciler.go:
```diff
@@ -173,8 +175,11 @@ func (rc *reconciler) reconcile() {
+			if err != nil {
+				glog.Errorf(mountedVolume.GenerateErrorDetailed(fmt.Sprintf("operationExecutor.NewVolumeHandler for UnmountVolume failed"), err).Error())
+			}
+			err = volumeHandler.UnmountVolumeHandler(mountedVolume.MountedVolume, rc.actualStateOfWorld)
```
Skip this if err?

jingxu97 commented on pkg/volume/util/operationexecutor/operation_generator.go:
```diff
@@ -627,6 +634,315 @@ func (og *operationGenerator) GenerateUnmountDeviceFunc(
+	}, blockVolumePlugin.GetPluginName(), nil
+}
+
+func (og *operationGenerator) GenerateUnmapDeviceFunc(
```
Maybe list the summary of the steps in the function comment?

jingxu97 commented on pkg/volume/util/operationexecutor/operation_generator.go:
```diff
@@ -821,3 +1137,23 @@ func checkNodeAffinity(og *operationGenerator, volumeToMount VolumeToMount, plug
+		(devicePathErr != nil && strings.Contains(devicePathErr.Error(), "does not exist")) {
+		// not a device path or path doesn't exist
+		//TODO: refer to #36092
+		glog.V(3).Infof("Not checking device path %s", deviceToDetach.DevicePath)
```
what does it mean by "not checking device path"? Could you make it more clear

msau42 commented on pkg/kubelet/volumemanager/populator/desired_state_of_world_populator.go:
```diff
@@ -381,6 +384,24 @@ func (dswp *desiredStateOfWorldPopulator) createVolumeSpec(
+			}
+			// Error if a container has volumeMounts but the volumeMode of PVC isn't Filesystem
+			// or if a container has volumeDevices but the volumeMode of PVC isn't Block
+			if (mountsMap[podVolume.Name] && volumeMode != v1.PersistentVolumeFilesystem) ||
```
Can you split the checks and error into two?  Then the error will be more specific.

mtanino replied:
ack.

msau42 commented on pkg/kubelet/volumemanager/reconciler/reconciler.go:
```diff
@@ -448,15 +482,51 @@ func (rc *reconciler) reconstructVolume(volume podVolume) (*reconstructedVolume,
+		// Check existence of mount point for filesystem volume or symbolic link for block volume
+		var isNotMount, islinkExist bool
+		var mountCheckErr, checkErr error
+		if attachablePlugin != nil && volume.volumeMode == v1.PersistentVolumeFilesystem {
```
Why do we need to check for filesystem mount point under the block volume feature?  Can this also be part of the handler?

mtanino replied:
ok. I think this check should be separated at block case and filesystem case.

mtanino replied:
Even if we enable block volume feature, both Filesystem and Block is coming to reconstruct. Therefore, we should check before calling mount point check or symlink existing check. Yes, I think we could add this handler.

mtanino replied:
done.

msau42 commented on pkg/kubelet/volumemanager/reconciler/reconciler.go:
```diff
@@ -574,33 +666,51 @@ func getVolumesFromPodDir(podDir string) ([]podVolume, error) {
+				// TODO: remove feature gate check after no longer needed
+				var volumeMode v1.PersistentVolumeMode
+				if utilfeature.DefaultFeatureGate.Enabled(features.BlockVolume) {
+					if volumesDir == path.Join(podDir, config.DefaultKubeletVolumeDevicesDirName) {
```
Can the mode be specified when defining the volumeDirs above?  I think it should be ok to specify v1.PersistentVolumeFilesystem outside of the feature gate in this particular scenario because it will be ignored when we're reading it.

mtanino replied:
I think we did this in previous. ack

msau42 commented on pkg/volume/util/operationexecutor/operation_generator.go:
```diff
@@ -627,6 +634,338 @@ func (og *operationGenerator) GenerateUnmountDeviceFunc(
+	}
+
+	return func() error {
+		// Try to unmap symlink on global map path
```
Should it be unmap pod link first, then global map?

mtanino replied:
yeah, it's reasonable.

msau42 commented on pkg/volume/util/operationexecutor/operation_generator.go:
```diff
@@ -627,6 +634,338 @@ func (og *operationGenerator) GenerateUnmountDeviceFunc(
+			return deviceToDetach.GenerateErrorDetailed("UnmapDevice.TearDownDevice failed", unmapErr)
+		}
+
+		// Plugin finished TearDownDevice(). Now globalMapPath is unnecessary, clean up it.
```
Should global map cleanup happen first, then teardown?  Otherwise your global map could have broken links?

mtanino replied:
@msau42 
No. TearDownDevice() may use plugin's stored data under globalMapPahth.
We should do TearDownDevice() first. Once tearDown is completed, stored data is unnecessary, no one uses the data, we can remove globalMapPath by removeall.

Also the globalPath is like "/var/lib/kubelet/plugins/kubernetes.io/fc/volumeDevices/target-lun-0", the globalPath is indicating dir not symbolic link.

msau42 commented on pkg/volume/util/operationexecutor/operation_generator.go:
```diff
@@ -821,3 +1160,23 @@ func checkNodeAffinity(og *operationGenerator, volumeToMount VolumeToMount, plug
+	} else {
+		deviceOpened, deviceOpenedErr = mounter.DeviceOpened(deviceToDetach.DevicePath)
+		if deviceOpenedErr != nil {
+			return false, deviceToDetach.GenerateErrorDetailed("UnmountDevice.DeviceOpened failed", deviceOpenedErr)
```
This "UnmountDevice" error is also going to return for block?

mtanino replied:
Should remove UnmountDevice. tag since this method is used both UnmountDevice and UnmapDevice.

msau42 commented on pkg/volume/util/util.go:
```diff
@@ -270,3 +278,201 @@ func stringToSet(str, delimiter string) (sets.String, error) {
+		return fmt.Errorf("Failed to remove map path. mapPath is empty")
+	}
+	glog.V(5).Infof("RemoveMapPath: mapPath %s", mapPath)
+	err := os.RemoveAll(mapPath)
```
Are you sure you want to remove all? In the normal successful case, there shouldn't be any content left in the directory right?

mtanino replied:
We call RemoveMapPath() following two cases.

(1) From GenerateUnmapVolumeFunc() -> remove only symlink under pod device map path 
(2) From GenerateUnmapDeviceFunc() -> remove whole global map path directory

When we call case (2), we already confirmed that;
- There is no reference from Pods to volume
- TearDownDevice is completed. Plugin's stored data under global map path directory are unnecessary at this point.

Therefore, I think we can global map path directory safely, otherwise the directory will be remained after volume is deleted.

mtanino commented:
@msau42 
I've addressed to all your previous review comments on "Volumemanager updates
c00b529". PTAL.

mtanino commented:
@msau42
I've addressed to all your previous review comments and added few updates. PTAL on Monday.
```
Volumemanager updates  … a2fcff7
- Fix/Add comments
- Change removing order of global map path and pod device map path
- Add CheckVolumeExistence volumeHandler
- Separate error cases for volumeMode and volumeMount/volumeDevices
  combination check
- Remove dupilicated symlink remove call under globalMapPath
- Update getVolumesFromPodDir logic
```

mtanino commented:
/retest

mtanino commented:
/retest

mtanino commented:
API was merged. These tests should be all green.

mtanino commented:
just rebased.

mtanino commented:
/retest

Commit:
generated code for Volumemanager change

mtanino commented:
/retest

mtanino commented:
/sig storage

mtanino commented:
/retest

mtanino commented:
pull-kubernetes-cross is hitting cross build failure. See following.
`fix cadvisor.New signature for cross build #56003`

jsafrane commented:
/approve
lgtm from my side, @msau42 ?

msau42 commented on pkg/kubelet/volumemanager/populator/desired_state_of_world_populator.go:
```diff
@@ -391,15 +391,22 @@ func (dswp *desiredStateOfWorldPopulator) createVolumeSpec(
+			if mountsMap[podVolume.Name] && volumeMode != v1.PersistentVolumeFilesystem {
 				return nil, "", fmt.Errorf(
-					"Invalid volumeMode. volumeMounts only supports 'Filesystem' mode and volumeDevices only supports 'Block' mode, pod: %q/%q, volume name: %v, volumeMode: %v",
+					"invalid combination of volumeMounts and volumeMode. volumeMounts expects volumeMode: 'Filesystem' but got '%v', pod: %q/%q, volume name: %v",
```
A simpler error message could be something like "Volume %q has volumeMode %q, but is specified in volumeMounts for pod %q"

mtanino replied:
ack.

mtanino replied:
@msau42 
I'd like to have namespace for purpose of troubleshooting. Updated format is like this.
```
Volume "dswp-test-volume-name" has volumeMode "Filesystem", but is specified in volumeDevices for pod "dswp-test"/"dswp-test-pod"
Volume "dswp-test-volume-name" has volumeMode "Block", but is specified in volumeMounts for pod "dswp-test"/"dswp-test-pod"
```

msau42 replied:
looks good, thanks!

msau42 commented on pkg/volume/util/operationexecutor/operation_executor.go:
```diff
@@ -994,6 +1015,22 @@ func (b BlockVolumeHandler) ReconstructVolumeHandler(_ volume.VolumePlugin, mapp
+// This method is handler for block volume
+func (b BlockVolumeHandler) CheckVolumeExistence(mountPath, volumeName string, mounter mount.Interface, uniqueVolumeName v1.UniqueVolumeName, podName volumetypes.UniquePodName, podUID types.UID, _ volume.AttachableVolumePlugin) (bool, error) {
+	blkutil := util.NewBlockVolumePathHandler()
+	if islinkExist, checkErr := blkutil.IsSymlinkExist(mountPath); checkErr != nil {
```
Can you add a comment explaining why the block volumes need to check for both attachable and non-attachable? (compared to filesystem, which only checks for attachable)

mtanino replied:
@msau42 
At the filesystem volume case, I believe that attachable check exists to avoid checking plugins like configmap who has directory under volumes but the directory isn't mounted.
```
e17c37b6-ce1b-11e7-aaf5-525400645a80/volumes/kubernetes.io~configmap
```

On the other hands, at the block volume case, either attachable or non-attachable, all of them should have pod device map path directory and symbolic link under the directory.
Therefore we don't need to check if attachable or not for block volume case.
```
./0e94706e-ca7b-11e7-b87a-525400645a80/volumeDevices/kubernetes.io~fc/block-pv001
```

msau42 commented:
lgtm, asking @saad-ali to take a final look

msau42 commented:
/kind feature
/priority important-soon

mtanino commented:
/retest

saad-ali commented on pkg/kubelet/volumemanager/reconciler/reconciler.go:
```diff
@@ -171,10 +174,12 @@ func (rc *reconciler) reconcile() {
-			glog.V(12).Infof(mountedVolume.GenerateMsgDetailed("Starting operationExecutor.UnmountVolume", ""))
-			err := rc.operationExecutor.UnmountVolume(
-				mountedVolume.MountedVolume, rc.actualStateOfWorld)
+			volumeHandler, err := operationexecutor.NewVolumeHandler(mountedVolume.VolumeSpec, rc.operationExecutor)
```
I don't understand this pattern. Why is a a `NewVolumeHandler` constructed before each usage?

mtanino replied:
`NewVolumeHandler` chooses a handler for filesystem or block volume based on `volumeMode` in mountedVolume.VolumeSpec. Since the volumeMode differs on each volume, we need to choose proper handler for a volume before each unmount volume operation.

saad-ali commented on pkg/volume/util/operationexecutor/operation_executor.go:
```diff
@@ -748,6 +842,200 @@ func (oe *operationExecutor) VerifyControllerAttachedVolume(
 }
 
+// VolumeStateHandler defines a set of operations for handling mount/unmount/detach/reconstruct volume-related operations
+type VolumeStateHandler interface {
```
How is this different then `OperationExecutor` itself and why is it needed?

mtanino replied:
This is wrapper to choose corresponding volume operations based on volumeMode.
To avoid adding `if volumeMode = "Block" {abc} else {xyz}` everywhere, we introduced this handler to select volumehandler from volumeSpec.

saad-ali commented on pkg/volume/util/operationexecutor/operation_executor.go:
```diff
@@ -105,6 +107,28 @@ type OperationExecutor interface {
+	// * Map volume to global map path using symbolic link.
+	// * Map the volume to the pod device map path using symbolic link.
+	// * Update actual state of world to reflect volume is mounted/mapped to the pod path.
+	MapVolume(waitForAttachTimeout time.Duration, volumeToMount VolumeToMount, actualStateOfWorld ActualStateOfWorldMounterUpdater) error
```
Why are new `MapVolume`/`UnmapVolume` operations needed. Why can't the exist `MountDevice`/`UnmountDevice`/MountVolume`/`UnmountVolume` could be extended to capture block behavior?

jsafrane replied:
`MountDevice` mounts in plugin global directory for reference counting. `MapVolume` does something completely different.

mtanino replied:
We have three operations `MountVolume`/`UnmountVolume`/`UnmountDevice` currently.
These operations execute `mount volume`, `unmount volume` and `detach volume` based on `mountpoint` and `mountpoint reference counting` information.
As for the Block Volume case, we uses `symbolic link` and `reference count using symbolic link` since the volume isn't formatted.

As @jsafrane mentioned, operations are completely different between filesystem and block, therefore we define new operations which indicates internal steps.

Filesystem | Purpose
-------------- | -------------
MountVolume | Call plugin's MountDevice() and SetUp()
UnmountVolume | Call plugin's TearDown()
UnmountDevice | Call plugin's UnmountDevice() based on reference count of mountpoint

Block | Purpose
-------------- | -------------
MapVolume | Call plugin's SetUpDevice() and create symlink to plugin dir and pod dir. Create loopbak for volume to avoid silent volume replacement
UnmapVolume | Delete symlink under plugin dir and pod dir
UnmapDevice | Call plugin's TearDownDevice() based on reference count of symlink. Delete loopback

saad-ali commented on pkg/volume/util/operationexecutor/operation_executor.go:
```diff
@@ -105,6 +107,28 @@ type OperationExecutor interface {
+	// If number of reference is zero, remove global map path directory and
+	// free a volume for detach.
+	// It then updates the actual state of the world to reflect that.
+	UnmapDevice(deviceToDetach AttachedVolume, actualStateOfWorld ActualStateOfWorldMounterUpdater, mounter mount.Interface) error
```
Why is this necessary? `MountDevice`/`UnmountDevice` do not ref count.

jsafrane replied:
We must have some way how to do reference counting to un-map a device when a pod is deleted when kubelet is down. We do the same with bind mounts for filesystems and scanning /proc/mounts.

mtanino replied:
Please see above matrix.

mtanino commented:
/retest

saad-ali reviewed (approved):
I don't like the way that `VolumeHandler` is structured, but, because this change is flag gated and is functional I'll approve. But please refactor this for beta.

saad-ali commented:
/lgtm
/approve

mtanino commented:
@saad-ali 

>>I don't like the way that VolumeHandler is structured, but, because this change is flag gated and is functional I'll approve. But please refactor this for beta.

I understand. I'll post this refactor for v1.10. Thank you for your review.

yujuhong commented:
/approve
